### PR TITLE
[tune] Tune experiment analysis improvements

### PR DESCRIPTION
**Pull request by krfricke**
## Why are these changes needed?

Depends on successful merge of #10627.

This PR adds useful utility functions to the ExperimentAnalysis object. Most functions were available before, but these wrapper properties allow easy access without specifying any parameters as long as `mode` and `metric` have been passed to `tune.run()`. 

It also introduces fetching pandas dataframes for the last trial results, a feature often found in other libraries.

The default lookup scope has been changed from `all` to `last`, as most users are interested comparing the final results, and not intermediate results.

## Related issue number

#10488

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

File: python/ray/tune/analysis/experiment_analysis.py [hunk 1]
```diff
@@ -316,7 +324,150 @@ def __init__(self,
...
+        if not self.default_metric or not self.default_mode:
+            raise ValueError(
+                "To fetch the `best_checkpoint`, pass a `metric` and `mode` "
+                "parameter to `tune.run()`. Alternatively, use the "
+                "`get_best_checkpoint(trial, metric, mode)` method to set the "
+                "metric and mode explicitly.")
+        best_trial = self.best_trial
+        return self.get_best_checkpoint(best_trial, self.default_metric,
+                                        self.default_mode)
+
```
Comment (krfricke) [inline python/ray/tune/analysis/experiment_analysis.py:382]: Currently this gets the best checkpoint of the best trial. Should it get the last checkpoint of the best trial instead?

Comment (richardliaw) [reply python/ray/tune/analysis/experiment_analysis.py:382]: no this sounds good.

File: python/ray/tune/analysis/experiment_analysis.py [hunk 2]
```diff
@@ -380,7 +531,7 @@ def get_best_trial(self, metric=None, mode=None, scope="all"):
                 "parameter?")
         return best_trial
 
-    def get_best_config(self, metric=None, mode=None, scope="all"):
+    def get_best_config(self, metric=None, mode=None, scope="last"):
```
Comment (richardliaw) [inline python/ray/tune/analysis/experiment_analysis.py:534]: nice - need to explicitly say this on the docs.

File: python/ray/tune/analysis/experiment_analysis.py [hunk 3]
```diff
@@ -316,7 +324,150 @@ def __init__(self,
...
+        if not self.default_metric or not self.default_mode:
+            raise ValueError(
+                "To fetch the `best_result`, pass a `metric` and `mode` "
+                "parameter to `tune.run()`. Alternatively, use "
+                "`get_best_trial(metric, mode).last_result` to set "
+                "the metric and mode explicitly and fetch the last result.")
+        return self.best_trial.last_result
+
+    @property
+    def best_result_df(self):
```
Comment (richardliaw) [inline python/ray/tune/analysis/experiment_analysis.py:438]: output type of these properties would be nice

File: python/ray/tune/analysis/experiment_analysis.py [hunk 4]
```diff
@@ -316,7 +324,150 @@ def __init__(self,
...
+
+    @property
+    def best_config(self):
+        """Get the config of the best trial of the experiment
+
+        The best trial is determined by comparing the last trial results
+        using the `metric` and `mode` parameters passed to `tune.run()`.
+
+        If you didn't pass these parameters, use
+        `get_best_config(metric, mode)` instead.
```
Comment (richardliaw) [inline python/ray/tune/analysis/experiment_analysis.py:353]: You probably want to also say something about `get_best_config(metric, mode, scope)`.

**Comment by richardliaw** [comment]
A couple tests failing - https://travis-ci.com/github/ray-project/ray/jobs/382378743

**Comment by krfricke** [comment]
I'll fix these once the metric/mode got merged

**Comment by krfricke** [comment]
Actually updated them now.

**Review by richardliaw — APPROVED** [review]

**Pull request merged**